components/suggest/src/yelp.rs

```diff
@@ -0,0 +1,350 @@
+/// parsing will be failed. Also, every words except Location needs to be registered in DB.
+/// Please refer to the query test in store.rs for all of combination.
+/// Currently, the maximum query length is determined while refering to having word lengths in DB
+/// and location names.
+/// max subject: 50 + pre-modifier: 10 + post-modifier: 10 + location-sign: 7 + location: 50 = 127 = 150.
+const MAX_QUERY_LENGTH: usize = 150;
+
+/// The max number of words consisting the modifier. To improve the SQL performance by matching with
+/// "keyword=:modifier" (please see is_modifier()), define this how many words we should check.
+const MAX_MODIFIER_WORDS_NUMBER: usize = 2;
+
+/// Insert Yelp suggestions into db.
+pub(crate) fn insert_suggestions(
+    conn: &Connection,
+    scope: &SqlInterruptScope,
+    record_id: &SuggestRecordId,
+    suggestion: &DownloadedYelpSuggestion,
+) -> Result<()> {
+    for keyword in &suggestion.subjects {
+        scope.err_if_interrupted()?;
+        conn.execute(
+            "INSERT INTO yelp_subjects(record_id, keyword) VALUES(:record_id, :keyword)",
+            named_params! {
+                ":record_id": record_id.as_str(),
+                ":keyword": keyword,
+            },
+        )?;
+    }
+
+    for keyword in &suggestion.pre_modifiers {
+        scope.err_if_interrupted()?;
+        conn.execute(
+            "INSERT INTO yelp_modifiers(record_id, type, keyword) VALUES(:record_id, :type, :keyword)",
+            named_params! {
+                ":record_id": record_id.as_str(),
+                ":type": Modifier::PreModifier,
+                ":keyword": keyword,
+            },
+        )?;
+    }
+
+    for keyword in &suggestion.post_modifiers {
+        scope.err_if_interrupted()?;
+        conn.execute(
+            "INSERT INTO yelp_modifiers(record_id, type, keyword) VALUES(:record_id, :type, :keyword)",
+            named_params! {
+                ":record_id": record_id.as_str(),
+                ":type": Modifier::PostModifier,
+                ":keyword": keyword,
+            },
+        )?;
+    }
+
+    for sign in &suggestion.location_signs {
+        scope.err_if_interrupted()?;
+        conn.execute(
+            "INSERT INTO yelp_location_signs(record_id, keyword, need_location) VALUES(:record_id, :keyword, :need_location)",
+            named_params! {
+                ":record_id": record_id.as_str(),
+                ":keyword": sign.keyword,
+                ":need_location": sign.need_location,
+            },
+        )?;
+    }
+
+    Ok(())
+}
+
+/// Fetch Yelp suggestion from given user's query.
+pub(crate) fn fetch_suggestion(
+    conn: &Connection,
+    query: &SuggestionQuery,
+) -> Result<Option<Suggestion>> {
+    if !query.providers.contains(&SuggestionProvider::Yelp) {
+        return Ok(None);
+    }
+
+    if query.keyword.len() > MAX_QUERY_LENGTH {
+        return Ok(None);
+    }
+
```

Comment:
Thanks!
Now, as we are keeping upper/lower case for the query of URL like `https://www.yelp.com/search?find_desc=Best+Ramen`, I wanted to pass the keyword as it is. 
However, it seems that the result of the URL is the same regardless of string cases. The only concern I have is the query form on Yelp. As it reflects the query as it is, might be better to do the same thing.
Please see the screenshot. What do you think?

<img width="697" alt="Screenshot 2024-01-09 at 11 07 22" src="https://github.com/mozilla/application-services/assets/1164542/e855281f-02a8-49cb-a59e-1ccc3c23cfc6">

Reply:
I'd like to hear your opinion as well, @0c0w3!

Reply:
Ah, it seems that we can do trimming here at least. I will do it.

Reply:
I'd be surprised if yelp handles queries case sensitively. If so, please disregard my suggestion :)

Reply:
Typically in urlbar we try to keep the user's casing and sometimes whitespace, that's true. I don't feel too strongly about it here. As Nan said, Yelp probably isn't case sensitive, so practically it shouldn't matter. If we need a tie breaker, I guess I would agree with Nan that we should normalize in one place since that would make the code a little simpler. I wouldn't expect most users to care about casing or even use uppercase when searching anyway.

Reply:
Yeah, the Yelp results don't seem to change by the string casing.
However, the string the user inputs into the Yelp form (like the screenshot) will remain in the same case on the result page. That's part of it, I thought that reflecting the cases user inputs to the form might be natural.
What do you think? If we don't have to care about it, I will change the code! Or, shall we consider this in a follow-up?

Reply:
I'm finding it hard to form an opinion either way -- I just don't think it's a big deal. There's a small perf cost to unnecessarily lowercasing the same strings more than once, but I would think it's miniscule overall, especially considering we force a limit on the length of the search string. Likewise there's a small amount of code inflation due to the multiple `to_lowercase()` calls. There's also a higher likelihood of introducing bugs due to case mismatches if we don't normalize early on in one place, like at the beginning of `fetch_suggestions()`. All these drawbacks seem minor to me.

If you want to preserve the case, I'm fine with that, but I'm not sure @ncloudioj is?

Reply:
It was just a nitpick for readability, not a blocker. Feel free to leave it as is. :)

I was wondering if those query normalization had special treatments in different places When I looked at the code. Looks like they all just do trimming and casing. So I suggested normalizing the original query once at the top to slightly reduce the cognitive load when reading those code paths in future.

Reply:
(Sorry for this long comment!)

One thing I do have a stronger opinion on, that I just thought of: The Yelp provider shouldn't bypass any query normalization that's done for every other provider.

Right now, `fetch_suggestions()` isn't really provider agnostic. i.e., it doesn't simply loop through a list of items that `impl` some generic `Provider` trait and call `fetch()` on each of them. So at the moment, we really don't have a clear implementation strategy where (1) normalization is done on the query and then (2) the normalized query is handed off to each provider. (Although https://github.com/mozilla/application-services/commit/f3cd74035a5eafcdd5925aadc2859ec59583ecdb did kind of start us down that path, which is good IMO.)

However, we've talked about rewriting `fetch_suggestions()` in the future so that it does work like that. And in that world, Yelp shouldn't be treated specially. It should get the same query as all the other providers, whether it's normalized or not.

I don't think we need to worry about this in this PR right now, but it's something to keep in mind if/when we rewrite `fetch_suggestions()`.